src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,861 @@
+        frame_bounds.start.n_preceding_rows().unwrap()
+    } else {
+        frame_bounds.end.n_following_rows().unwrap()
+    };
+
+    for _ in 0..n_rows_to_move {
+        if LEFT {
+            cursor.move_prev();
+        } else {
+            cursor.move_next();
+        }
+        if cursor.position().is_ghost() {
+            break;
+        }
+    }
+    cursor
+        .key()
+        .or_else(|| {
+            if LEFT {
+                part_with_delta.first_key()
+            } else {
+                part_with_delta.last_key()
+            }
+        })
+        .unwrap()
+}
+
+#[cfg(test)]
+mod tests {
+    use risingwave_expr::window_function::{FrameBound, RowsFrameBounds};
+
+    use super::*;
+
+    #[test]
+    fn test_merge_rows_frame() {
+        fn assert_equivalent(bounds1: RowsFrameBounds, bounds2: RowsFrameBounds) {
+            assert_eq!(bounds1.start.to_offset(), bounds2.start.to_offset());
+            assert_eq!(bounds1.end.to_offset(), bounds2.end.to_offset());
+        }
+
+        assert_equivalent(
+            merge_rows_frames(&[]),
+            RowsFrameBounds {
+                start: FrameBound::CurrentRow,
+                end: FrameBound::CurrentRow,
+            },
+        );
+
+        let frames = [
+            &RowsFrameBounds {
+                start: FrameBound::Preceding(3),
+                end: FrameBound::Preceding(2),
+            },
+            &RowsFrameBounds {
+                start: FrameBound::Preceding(1),
+                end: FrameBound::Preceding(4),
+            },
+        ];
+        assert_equivalent(
+            merge_rows_frames(&frames),
+            RowsFrameBounds {
+                start: FrameBound::Preceding(4),
+                end: FrameBound::CurrentRow,
+            },
+        );
+
+        let frames = [
+            &RowsFrameBounds {
+                start: FrameBound::Preceding(3),
+                end: FrameBound::Following(2),
+            },
+            &RowsFrameBounds {
+                start: FrameBound::Preceding(2),
+                end: FrameBound::Following(3),
+            },
+        ];
+        assert_equivalent(
+            merge_rows_frames(&frames),
+            RowsFrameBounds {
+                start: FrameBound::Preceding(3),
+                end: FrameBound::Following(3),
+            },
+        );
+
+        let frames = [
+            &RowsFrameBounds {
+                start: FrameBound::UnboundedPreceding,
+                end: FrameBound::Following(2),
+            },
+            &RowsFrameBounds {
+                start: FrameBound::Preceding(2),
+                end: FrameBound::UnboundedFollowing,
+            },
+        ];
+        assert_equivalent(
+            merge_rows_frames(&frames),
+            RowsFrameBounds {
+                start: FrameBound::UnboundedPreceding,
+                end: FrameBound::UnboundedFollowing,
+            },
+        );
+    }
+
+    mod rows_frame_tests {
```

Comment:
Unit tests in this module are migrated from previous `find_affected_ranges_tests`

Reply:
Do we have `find_affected_ranges` tests? I could not ctrl-F the function call here. Looks like the tests were deleted.